app/javascript/lib/scatter-plot.js

```diff
@@ -0,0 +1,308 @@
+function getScatterPlotLayout(rawPlot) {
+  const { numRows, numColumns } = rawPlot
+  const { height, width } = calculatePlotRect(numRows, numColumns)
+  let layout = getBaseLayout(height, width)
+
+  let dimensionProps
+  if (rawPlot.is3D) {
+    const camera = $('#cluster-plot').data('camera')
+    dimensionProps = get3DScatterProps(camera, rawPlot)
+    layout.scene = dimensionProps
+  } else {
+    dimensionProps = get2DScatterProps(rawPlot)
+    layout = Object.assign(layout, dimensionProps)
+  }
+
+  return layout
+}
+
+/** Renders a Plotly scatter plot for "Clusters" tab */
+function renderScatterPlot(rawPlot, plotId, legendId) {
+  let { data } = rawPlot
+
+  const layout = getScatterPlotLayout(rawPlot)
+
+  if (rawPlot.annotParams.type === 'group' && !rawPlot.gene) {
+    data = setMarkerColors(data)
+  }
+
+  Plotly.newPlot(plotId, data, layout)
+
+  if (legendId) {
+    $(`#${legendId}`).html(
+      `<p class="text-center help-block">${rawPlot.description}</p>`
+    )
+  }
+
+  // access actual target div, not jQuery object wrapper for relayout event
+  const scatterPlotDiv = document.getElementById(plotId)
+  scatterPlotDiv.on('plotly_relayout', cameraData => {
+    if (typeof cameraData['scene.camera'] !== 'undefined') {
+      const newCamera = cameraData['scene.camera']
+      $(`#${plotId}`).data('camera', newCamera)
+    }
+  })
+}
+
+/**
+ * Load and draw scatter plot.  Handles clusters and spatial groups, numeric
+ * and group annotations.
+ *
+ * @param {Object} clusterParams Parameters for `/clusters` plot API endpoint
+ * @param {Object} frame Plot UI properties not from plot API data
+ */
+export async function scatterPlot(clusterParams, frame) {
+  // Copy by value; preserves properties (e.g. plotId) across `await`
+  frame = Object.assign({}, frame)
+
+  const { plotId, hasLegend } = frame
+  console.log(plotId)
+
+  const legendId = (hasLegend ? `${plotId}-legend` : null)
+  const legendHtml = (hasLegend ? `<div id="${legendId}"></div>` : '')
+
+  $(frame.selector).append(
+    `<div id="${plotId}" class="plot"></div>
+      ${legendHtml}
+    </div>`
+  )
+
+  // Set default number of rows and columns of plots
+  if (!frame.numRows) {frame.numRows = 1}
+  if (!frame.numColumns) {frame.numColumns = 1}
+
+  const $plotElement = $(`#${plotId}`)
+  const spinnerTarget = $plotElement[0]
+  const spinner = new Spinner(window.opts).spin(spinnerTarget)
+  // $plotElement.data('spinner', spinner)
+
+  const { accession, cluster, annotation, subsample, gene } = clusterParams
+  let isAnnotatedScatter = null
+  if (
+    'isAnnotatedScatter' in clusterParams && clusterParams.isAnnotatedScatter
+  ) {
+    isAnnotatedScatter = true
+  }
+
+  $('#search_annotation').val(annotation)
+  $('#gene_set_annotation').val(annotation)
+
+  let rawPlot = await fetchCluster(
+    accession, cluster, annotation, subsample, null, gene, isAnnotatedScatter
+  )
+  rawPlot = Object.assign(rawPlot, frame)
```

Comment:
wait, the comment above says the frame is not related to the cluster opts, but now you're merging them?  And maybe rawPlot shoudl be named 'plotConfig' or 'plotOpts' since it's not an actual plot object or container

Reply:
To help clarify (or at least not mislead), I renamed `frame` to `props`, and renamed the return variable from `fetchCluster` to `fetchedData`.